### PR TITLE
Introduction of cart workflow

### PR DESCRIPTION
**Pull request by 6R1M5L07H**
The feature includes two models: cart and cart item. The cart is created as an intermediate step between the selection of the goods so that a customer can collect items in said cart as cart items and check them out in one step.

Resolves request #69

### CODE REVIEW

File: l10n/en.json [hunk 1]
```diff
@@ -6,6 +6,17 @@
   "help": "\uD83D\uDE80 Help",
   "help_string": "<b>Support</b>",
   "help_button": "Admin",
+  "cart": "\u1F6D2 Cart",
+  "cart_item_button": "{cart_item_subcategory_name} ({cart_item_quantity}) = {cart_item_total}{cart_item_currency} \uD83D\uDDD1",
+  "cart_line_item": "{cart_item_subcategory_name} ({cart_item_quantity!s}) = {cart_item_total:.2f}{cart_item_currency!s} \n",
```
Comment (ilyarolf) [inline l10n/en.json:11]: Apparently these strings represent the same thing, maybe we should use the same one?
I would like the buttons in the cart to look exactly the same as the buttons in the screenshot, that way the user interface will look uniform.
image

File: models/buyItem.py [hunk 2]
```diff
@@ -8,7 +8,12 @@ class BuyItem(Base):
     __tablename__ = "buyItem"
 
     id = Column(Integer, primary_key=True, unique=True, nullable=False)
+    buy_id = Column(Integer, ForeignKey("buys.id"), nullable=False)
+    buy = relationship("Buy", backref="buys")
+    item_id = Column(Integer, ForeignKey("items.id"), nullable=False)
+    item = relationship("Item", backref="items")
```
Comment (ilyarolf) [inline models/buyItem.py:14]: Why are you duplicating something that's already there?
Roll back the changes.

File: models/cart.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+    id = Column(Integer, primary_key=True)
+    cart_id = Column(Integer, ForeignKey("carts.id"), nullable=False)
+    cart: Mapped["Cart"] = relationship(back_populates="cart_items")
+    category_id: Mapped[int] = mapped_column(ForeignKey("categories.id"))
+    category_name: Mapped[str] = mapped_column(default=None, nullable=True)
+    subcategory_id: Mapped[int] = mapped_column(ForeignKey("subcategories.id"), passive_deletes="all")
+    subcategory_name: Mapped[str] = mapped_column(default=None, nullable=True)
+    quantity = Column(Integer, default=0)
+    a_piece_price = Column(Float, default=0.0)
```
Comment (ilyarolf) [inline models/cart.py:38]: 1) Don't use mapped_column, look at what other models look like and redesign those models to look like them.
2) The CartItem model should be moved to a separate file like I did with BuyItem
3) It is correct to use foreign key only to the primary key of another table, in this case using telegram_id is not correct, even if it works.
4) The cart_items field as I understand it is not even created in the database, and judging by the code I don't quite understand why it is needed. 
5) Why do you store category_name and subcategory_name if it can be obtained via category_id and subcategory_id?
6) The “a_piece_price” field can be retrieved by querying the items table by category_id and subcategory_id.

File: models/currency.py
Comment (ilyarolf) [inline models/currency.py:1]: If you decide to add CURRENCY as a field in the config, you need to implement its use everywhere, both in L10n and crypto_api.

File: models/item.py
Comment (ilyarolf) [inline models/item.py:1]: It seems that nothing much has changed in this file, if so, roll back the changes to make it easier for me to check.

File: models/shipment.py
Comment (ilyarolf) [inline models/shipment.py:1]: Shipment is not used in this code, delete this file.
Even if it will be used in the future, it is not needed now.

File: readme.md
Comment (ilyarolf) [inline readme.md:1]: Return the original readme.md

File: run.py [hunk 4]
```diff
@@ -20,14 +22,24 @@
...
     all_categories_button = types.KeyboardButton(text=Localizator.get_text_from_key("all_categories"))
     my_profile_button = types.KeyboardButton(text=Localizator.get_text_from_key("my_profile"))
     faq_button = types.KeyboardButton(text=Localizator.get_text_from_key("faq"))
     help_button = types.KeyboardButton(text=Localizator.get_text_from_key("help"))
-    keyboard = [[all_categories_button, my_profile_button], [faq_button, help_button]]
+    checkout_button = types.KeyboardButton(text=Localizator.get_text_from_key("checkout"))
+
+    # TODO: insert item count here, these code lines are not correct yet
+    #product_count = await CartService.get_all_cart_items_count(user_telegram_id)
+    #cart_button_label = Localizator.get_text_from_key("cart") + " (" + str(product_count) + ")"
```
Comment (ilyarolf) [inline run.py:36]: The final code must not contain commented lines of code

File: run.py [hunk 5]
```diff
@@ -1,15 +1,17 @@
...
-from config import SUPPORT_LINK
-import logging
 from bot import dp, main
-from multibot import main as main_multibot
+from config import SUPPORT_LINK
 from handlers.admin.admin import admin_router
 from handlers.user.all_categories import all_categories_router
+from handlers.user.cart import cart_router
 from handlers.user.my_profile import my_profile_router
+from multibot import main as main_multibot
```
Comment (ilyarolf) [inline run.py:14]: Why do you move lines up and down unnecessarily?

File: run.py [hunk 6]
```diff
@@ -20,14 +22,24 @@
 
 @main_router.message(Command(commands=["start", "help"]))
 async def start(message: types.message):
+    user_telegram_id = message.chat.id
+    user_telegram_username = message.from_user.username
+
```
Comment (ilyarolf) [inline run.py:27]: Why do you move lines up and down unnecessarily?

File: services/category.py [hunk 7]
```diff
@@ -67,3 +68,13 @@ async def get_maximum_page():
...
                 return math.trunc(max_page / CategoryService.items_per_page)
 
+    @staticmethod
+    async def get_name(category_id) -> str:
+        async with async_session_maker() as session:
+            stmt = select(Category.name).where(Category.id == category_id)
+            category_name = await session.execute(stmt)
+            category_name = category_name.scalar()
+            if category_name is None:
+                return ""
```
Comment (ilyarolf) [inline services/category.py:78]: return None

File: services/shipment.py
Comment (ilyarolf) [inline services/shipment.py:1]: Delete if not in use

File: services/subcategory.py [hunk 8]
```diff
@@ -66,3 +66,13 @@ async def delete_if_not_used(subcategory_id: int):
...
                 await session.commit()
+
+    @staticmethod
+    async def get_name(subcategory_id) -> str:
+        async with async_session_maker() as session:
+            stmt = select(Subcategory.name).where(Subcategory.id == subcategory_id)
+            subcategory_name = await session.execute(stmt)
+            subcategory_name = subcategory_name.scalar()
+            if subcategory_name is None:
+                return ""
```
Comment (ilyarolf) [inline services/subcategory.py:77]: return None

Comment (ilyarolf) [reply services/subcategory.py:77]: subcategory = SubcategoryService.get_by_primary_key(subcategory_id)
subcategory_name = subcategory.name

File: services/category.py (hunk 7 above)
Comment (ilyarolf) [reply services/category.py:78]: category = CategoryService.get_by_primary_key(subcategory_id)
category_name = category.name

File: services/user.py [hunk 9]
```diff
@@ -155,7 +155,9 @@ async def update_top_up_amount(telegram_id, deposit_amount):
     async def is_buy_possible(telegram_id, total_price):
         user = await UserService.get_by_tgid(telegram_id)
         balance = user.top_up_amount - user.consume_records
-        return balance >= total_price
+        #TODO: change back
+        return True
+        #return balance >= total_price
```
Comment (ilyarolf) [inline services/user.py]: Fix it

File: utils/notification_manager.py [hunk 10]
```diff
@@ -77,22 +80,35 @@ async def new_deposit(new_crypto_balances: dict, deposit_amount_usd, telegram_id
         await NotificationManager.send_to_admins(message, user_button, bot)
 
     @staticmethod
-    async def new_buy(subcategory_id: int, quantity: int, total_price: float, user: User, bot):
-        subcategory = await SubcategoryService.get_by_primary_key(subcategory_id)
-        message = ""
+    #async def new_buy(subcategory_id: int, quantity: int, total_price: float, user: User, bot):
```
Comment (ilyarolf) [inline utils/notification_manager.py:83]: remove commentary

File: utils/notification_manager.py [hunk 11]
```diff
@@ -77,22 +80,35 @@ async def new_deposit(new_crypto_balances: dict, deposit_amount_usd, telegram_id
...
+                    subcategory_name=subcategory.name) + "\n"
+            else:
+                message += Localizator.get_text_from_key("new_purchase_notification_with_username").format(
+                    telegram_id=telegram_id,
+                    total_price=cart_item_total,
+                    quantity=cart_item.quantity,
+                    subcategory_name=subcategory.name) + "\n"
+        dashes = "-" * min(config.MAX_LINE_WITH, line_max_width)
+        message += dashes + "\n"
+        message += Localizator.get_text_from_key("cart_grand_total_string").format(cart_grand_total=cart_grand_total, cart_item_currency=config.CURRENCY)
```
Comment (ilyarolf) [inline utils/notification_manager.py:113]: To be honest, I don't really like those dashes.
You can use Telegram styles to make the final message more beautiful.

https://core.telegram.org/bots/api#html-style

<b>bold</b>, <strong>bold</strong>
<i>italic</i>, <em>italic</em>
<u>underline</u>, <ins>underline</ins>
<s>strikethrough</s>, <strike>strikethrough</strike>, <del>strikethrough</del>
<span class="tg-spoiler">spoiler</span>, <tg-spoiler>spoiler</tg-spoiler>
<b>bold <i>italic bold <s>italic bold strikethrough <span class="tg-spoiler">italic bold strikethrough spoiler</span></s> <u>underline italic bold</u></i> bold</b>
<a href="http://www.example.com/">inline URL</a>
<a href="tg://user?id=123456789">inline mention of a user</a>
<tg-emoji emoji-id="5368324170671202286">👍</tg-emoji>
<code>inline fixed-width code</code>
<pre>pre-formatted fixed-width code block</pre>
<pre><code class="language-python">pre-formatted fixed-width code block written in the Python programming language</code></pre>
<blockquote>Block quotation started\nBlock quotation continued\nThe last line of the block quotation</blockquote>
<blockquote expandable>Expandable block quotation started\nExpandable block quotation continued\nExpandable block quotation continued\nHidden by default part of the block quotation started\nExpandable block quotation continued\nThe last line of the block quotation</blockquote>

File: services/item.py [hunk 12]
```diff
@@ -67,10 +68,12 @@ async def get_items_by_buy_id(buy_id: int) -> list:
...
     async def get_unsold_subcategories_by_category(category_id: int, page) -> list[Item]:
-        async with async_session_maker() as session:
-            stmt = select(Item).join(Subcategory, Subcategory.id == Item.subcategory_id).where(
+        async with (async_session_maker() as session):
+            stmt = (select(Item, Category.name, Subcategory.name)
+                    .join(Subcategory, Subcategory.id == Item.subcategory_id)
+                    .join(Category, Category.id == category_id).where(
                 Item.category_id == category_id, Item.is_sold == 0).group_by(Subcategory.name).limit(
-                ItemService.items_per_page).offset(ItemService.items_per_page * page)
+                ItemService.items_per_page).offset(ItemService.items_per_page * page))
```
Comment (ilyarolf) [inline services/item.py:76]: Why did you redo the query if in create_subcategory_buttons, item.subcategory.name and item.category.name are not used?
As far as I know joinedload is disabled here.

File: services/cart.py [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+                await CartService.create_cart_item(
+                    cart_item=CartItem(
+                           cart=cart
+                         , cart_id=cart.id
+                         , category_id=cart_item.category_id
+                         , category_name=cart_item.category_name
+                         , subcategory_id=cart_item.subcategory_id
+                         , subcategory_name=cart_item.subcategory_name
+                         , quantity=cart_item.quantity
+                         , a_piece_price=cart_item.a_piece_price))
```
Comment (ilyarolf) [inline services/cart.py:97]: Commas are better placed at the end rather than the beginning.

File: services/cart.py [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+            await db_session.delete(cart_item)
+            await db_session.commit()
+
+
+    @staticmethod
+    async def add_to_cart(cart_item: CartItem, cart: Cart):
+
+        async with (async_session_maker() as db_session):
+            # if there exists a cart with a cart_item with the same category and subcategory, increase the quantity
+            # and if not, there is either no cart_item in the cart at all or no cart_item of the same (sub-)category
```
Comment (ilyarolf) [inline services/cart.py:78]: Delete the comments

File: services/buyItem.py
Comment (ilyarolf) [inline services/buyItem.py:1]: This kind of change is better not to commit, it makes it harder to watch your PR. A lot of cognitive load.

File: run.py [hunk 15]
```diff
@@ -50,10 +61,11 @@ async def support(message: types.message):
     admin_keyboard_builder.button(text=Localizator.get_text_from_key("help_button"), url=SUPPORT_LINK)
     await message.answer(Localizator.get_text_from_key("help_string"), reply_markup=admin_keyboard_builder.as_markup())
 
-
 main_router.include_router(admin_router)
 main_router.include_router(my_profile_router)
 main_router.include_router(all_categories_router)
+#main_router.include_router(checkout_router)
```
Comment (ilyarolf) [inline run.py:67]: The final code must not contain commented lines of code

File: config.py
Comment (ilyarolf) [inline config.py:1]: From the MAX_LINE_WITH variable, it has no meaning as a configuration variable.
If you want to use CURRENCY as a variable, you need to add changes to L10n, and to crypto_api. Right now this variable is non-functional, and it has no meaning either.

**Review by ilyarolf — CHANGES_REQUESTED** [review]
There's still a lot of work to be done before pouring in that PR.
General Comments:
1) Use an approach that is already in use at the moment:
- Commas at the beginning instead of at the end;
- Distinguish models for SQLAlchemy from previous models;
- Products in cart can be shown the same way as Subcategories.
All this is unnecessary and makes the code heterogeneous.
2) There should not be any comments in the files, only TODO is allowed.
3) Do not add useless files that are not used anywhere, even if you plan to use the code from these files later, they are of no use now.
4) Don't create formatting changes, this causes the file to show up as changed, even though the actual essence of the file hasn't changed. This increases the cognitive load for the PR checker.

File: .env
[the remark this answers is not in the document]
Comment (6R1M5L07H) [reply .env:1]: added one paramater for max line width and another one to use other fiat currencies than USD. Rest rolled back

File: .env.template
[the remark this answers is not in the document]
Comment (6R1M5L07H) [reply .env.template:1]: If you want to test locally, .env-file-values need to be filled. In other projects I only saw a template-file and no .env-file which you should only have locally and not in the repo. In its current configuration, you have to overwrite .env for local testing and have to make sure before every commit to not add your data to the repo

File: docker-compose.yml
[the remark this answers is not in the document]
Comment (6R1M5L07H) [reply docker-compose.yml:1]: done

[the remark this answers is not in the document]
Comment (6R1M5L07H) [reply docker-compose.yml:1]: I added the two additional .env-parameters

**Comment by 6R1M5L07H** [comment]
>     * Commas at the beginning instead of at the end;

Done whenever there is a line being split over several lines. Would you change the existing classes accordingly?

>     * Distinguish models for SQLAlchemy from previous models;
Please elaborate

>     * Products in cart can be shown the same way as Subcategories.
>       All this is unnecessary and makes the code heterogeneous.

IMO these are two separate entities. If any, the CartItem resembles more the Buy Entity but still I think that you need a different entity. Here is why:
- Subcategory (which I would rephrase as "Article" tbh, but that is something different) is something static in the DB and is independent from a User's class or a Buy, so it exists from my understanding to represent the Article Catalogue
- Buy is bound to the User and exists for the purpose of having a history
- CartItem however is meant to be volatile since it only exists while the Cart exists and the Cart only exists until its content - the CartItems - are transformed into Buy objects. The CartItem is imo needed to implement the payment process on a "pay as you go" approach instead of the existing deposit approach

Happy to discuss

>     2. There should not be any comments in the files, only TODO is allowed.
I'll follow your direction, although I think that comments don't hurt anyone. I had to understand your way of thinking for some time and comments would have helped me a lot. If you change your point of view here, let me know, please. Until then, I remove the comments (I do not mean commented code but actual comments btw)

>     3. Do not add useless files that are not used anywhere, even if you plan to use the code from these files later, they are of no use now.

Agreed and removed.

>     4. Don't create formatting changes, this causes the file to show up as changed, even though the actual essence of the file hasn't changed. This increases the cognitive load for the PR checker.

Agreed, will improve myself here, thank you

File: handlers/user/all_categories.py [hunk 16]
```diff
@@ -7,39 +7,47 @@
...
+                                 category_id=category_id,
+                                 category_name=category_name,
+                                 subcategory_id=subcategory_id,
+                                 subcategory_name=subcategory_name,
+                                 price=price,
                                  total_price=total_price,
-                                 quantity=quantity, confirmation=confirmation, page=page).pack()
+                                 quantity=quantity,
+                                 confirmation=confirmation,
+                                 page=page).pack()
```
[the remark this answers is not in the document]
Comment (6R1M5L07H) [reply handlers/user/all_categories.py:50]: Good point, changed accordingly

File: handlers/user/all_categories.py [hunk 17]
```diff
@@ -91,7 +105,7 @@ async def all_categories(message: Union[Message, CallbackQuery]):
             category_inline_buttons = await add_pagination_buttons(category_inline_buttons, zero_level_callback,
                                                                    CategoryService.get_maximum_page(),
                                                                    AllCategoriesCallback.unpack, None)
-            await message.answer(Localizator.get_text_from_key("all_categories"), parse_mode=ParseMode.HTML,
+            await message.answer(Localizator.get_text_from_key("cart"), parse_mode=ParseMode.HTML,
```
[the remark this answers is not in the document]
Comment (6R1M5L07H) [reply handlers/user/all_categories.py:108]: Solved

File: services/cart.py (hunk 14 above)
Comment (6R1M5L07H) [reply services/cart.py:78]: see my comment above - happy to discuss

File: config.py
Comment (6R1M5L07H) [reply config.py:1]: MAX_LINE_WIDTH: the reason for this variable is that the dashes below the CartItems, which separate the items from the grand total, should not span over multiple lines. I thought that it might make sense to have this configurable. If not, I can configure it as a constant. What would you prefer?

The CURRENCY variable however is something I wanted to propose to introduce to be able to offer other fiat currency than only USD

File: l10n/en.json (hunk 1 above)
Comment (6R1M5L07H) [reply l10n/en.json:11]: It should be clear that every CartItem represented by a button has a function, which is that you can delete it from the cart when you tap on it. I haven't had a better idea on how to represent that which is why I added the trashbin emoticon at the end of the cart_item_button in contrast to the cart_line_item, which is only used after the purchase confirmation when the customer cannot remove anything from the cart anymore.

File: models/buyItem.py (hunk 2 above)
Comment (6R1M5L07H) [reply models/buyItem.py]: done

File: models/cart.py (hunk 3 above)
Comment (6R1M5L07H) [reply models/cart.py:38]: @1: maybe my understanding is wrong, but from what I learnt from [here](https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#declarative-table-with-mapped-column) is that the way that I'm defining the models not only works the way you defined your models but in a declarative way, plus "mapped_column() supersedes the use of Column()"

@2: true

@3: that made me thinking. First, I agree that per definition an FK is a reference to another table's PK, but as long as the refererred column is unique, from a technical point an FK constraint is okay. But I see your point and want to let you know my idea here: if I would refer the user by its ID, I would have to make another DB query to first get the user.ID from the telegram ID because the telegram ID is already contained in the message object. The message does of course not contain the user.id field. I think that this extra DB query is unneccessary so I wanted to use what's already there. I can redesign it if you want.

@4: I was sticking to the documentation from said documentation of [declarative definition](https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#declarative-table-with-mapped-column)

@5: to reduce DB queries. From my experience, sometimes the de-normalization of DB schemas is useful when you see an impact on the performance. Let's picture a shop where you would accumulate hundrets of CartItems in one purchase. Since the (sub-)category names are listed in the confirm message, I would have to query the DB once for every CartItem in the Cart. I think that this bit of redundancy would be better for the performance and not increase storage a lot, because the CartItem is meant to be sort of short-lived

@6: same as 5

Again: happy to discuss

File: models/currency.py
Comment (6R1M5L07H) [reply models/currency.py:1]: true. I can either remove it for now or keep it for later changes. I'll follow your preference

File: models/item.py
Comment (6R1M5L07H) [reply models/item.py:1]: true, there is only one comment I added because I needed some time to understand, what the Item entity was about

File: run.py (hunk 4 above)
Comment (6R1M5L07H) [reply run.py:36]: I'll leave the TODO in if you agree

File: run.py (hunk 5 above)
Comment (6R1M5L07H) [reply run.py:14]: as far as I can tell this was done by my IDE. okay for now to leave it?

File: run.py (hunk 6 above)
Comment (6R1M5L07H) [reply run.py:27]: see previous comment

File: services/category.py (hunk 7 above)
Comment (6R1M5L07H) [reply services/category.py:78]: method returns str which is not None

File: services/subcategory.py (hunk 8 above)
Comment (6R1M5L07H) [reply services/subcategory.py:77]: thats better

**Comment by ilyarolf** [comment]
@6R1M5L07H 
1) About SQLAlchemy models.
image
image
Use a model field format similar to Users, i.e.
`Column({type}, nullable={value}, etc...)`
2)

> Please elaborate
> 
>> * Products in cart can be shown the same way as Subcategories.
>>   All this is unnecessary and makes the code heterogeneous.

What was meant here is that you need to rework the way you describe the model fields. What I described in the 1st paragraph with screenshots. 

3) About the comments in the code. Try to use inline comments only in complex cases, for example in the [db.py](https://github.com/ilyarolf/AiogramShopBot/blob/develop/db.py) file it is described in detail why it is necessary to import SQLAlchemy models even though they are not used in this file, and also it is described why SQLCipher will not work on Windows.
Comments that I will always be glad to welcome are [PEP8 docstrings](https://www.mybluelinux.com/python-docstrings-annotations-and-pep8/).

**Review by ilyarolf — APPROVED** [review]
